network/transport/grpc/connection.go

```diff
@@ -160,24 +211,34 @@ func (mc *conn) open(tlsConfig *tls.Config, connectedCallback func(grpcConn *grp
 		return
 	}
 
-	mc.closers = nil
-
 	mc.connector = createOutboundConnector(mc.getPeer().Address, mc.dialer, tlsConfig, func() bool {
-		return !mc.connected()
+		return !mc.connected(AnyProtocol)
 	}, func(conn *grpc.ClientConn) {
 		mc.mux.Lock()
 		mc.grpcOutboundConnection = conn
 		mc.mux.Unlock()
 
 		connectedCallback(conn)
 	})
-	go mc.connector.loopConnect()
+	mc.connector.start()
 }
 
-func (mc *conn) connected() bool {
-	mc.mux.Lock()
-	defer mc.mux.Unlock()
+func (mc *conn) connected(protocol string) bool {
+	mc.mux.RLock()
+	defer mc.mux.RUnlock()
+
+	if protocol != AnyProtocol {
+		// Check for specific protocol
+		if mc.grpcOutboundStreams[protocol] != nil {
+			return true
+		}
+		if mc.grpcInboundStreams[protocol] != nil {
+			return true
+		}
```

Comment:
This could panic right? Shouldn't we use `if _, ok := mc.grpcOutboundStreams[protocol]; ok` here?

Reply:
Reading a non-existing item from a map returns either `nil` (in case of an interface type) or the default value (in case of a value type). `ok` allows you to differentiate with the entry being `nil`/default or non-existent. The value here is an interface, so it just returns nil. Proof:

```go
func Test_t(t *testing.T) {
	println(fmt.Sprintf("%v", map[string]interface{}{"foo": "bar"}["fuu"]))
}
```

```
=== RUN   Test_t
<nil>
--- PASS: Test_t (0.00s)
```

Reply:
Wow. I never knew... Thanks 😅